dev/fattest.simplicity/src/componenttest/containers/ArtifactoryImageNameSubstitutor.java

```diff
@@ -0,0 +1,84 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package componenttest.containers;
+
+import org.testcontainers.utility.DockerImageName;
+import org.testcontainers.utility.ImageNameSubstitutor;
+
+import com.ibm.websphere.simplicity.log.Log;
+
+import componenttest.topology.utils.ExternalTestService;
+
+/**
+ * An image name substituter is configured in testcontainers.properties and will transform docker image names.
+ * Here we use it to apply a private registry prefix so that in remote builds we use an internal mirror
+ * of Docker Hub, instead of downloading from Docker Hub in each build which causes rate limiting issues.
+ */
+public class ArtifactoryImageNameSubstitutor extends ImageNameSubstitutor {
```

Comment:
I do not really understand what this class is doing.  Or more accurately how our infrastructure is set up to support this.  

Do we have a system set up that caches docker images from docker hub and hosts them within our test environment? 

Will developers running tests locally still be able to run against our remote docker machines?

Reply:
The image name substitutor is an SPI that Testcontainers recently added specifically for addressing the docker hub rate limiting issue:
https://www.testcontainers.org/features/image_name_substitution/

Basically you can configure an `ImageNameSubstitutor` globally (as we do in this PR) and it will remap all of the docker image names used by Testcontainers. So if some FAT bucket does:
```java
public static GenericContainer myContainer = new GenericContainer("someImage:1.0");
```
When the substitutor is activated the code will _effectively_ be:
```java
public static GenericContainer myContainer = new GenericContainer("com.mycompany.registry/someImage:1.0");
```
Where `com.mycompany.registry` is a companies private docker registry (ideally a mirror of Docker Hub). I've set up a mirror registry within the IBM network to handle this. A registry mirror is basically a lazily-populated cache. If a bucket does a pull of a new image `someImage:1.0`, the mirror will check to see if it's available within the mirror, and if not it'll download the image from Docker Hub and cache it for future use. So that way the next time we do a pull of `someImage:1.0` it comes from the mirror registry and we don't need to touch Docker Hub, thus avoiding the rate limiting issue.

> Will developers running tests locally still be able to run against our remote docker machines?

Yes. If we detect that remote docker is being used, then the new `ExternalTestServiceDockerClient.setupTestcontainers` method will rewrite the `.testcontainers.properties` file to use the `ArtifactoryImageNameSubstitutor` and it will generate a `~/.docker/config.json` with a the private docker registry mirror auth token.

Reply:
Okay thanks for that extra explanation! 👍